internal/rukpakctl/upload.go

```diff
@@ -0,0 +1,122 @@
+package rukpakctl
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/rest"
+
+	"github.com/operator-framework/rukpak/internal/util"
+)
+
+// BundleUploader uploads bundle filesystems to rukpak's upload service.
+type BundleUploader struct {
+	UploadServiceName      string
+	UploadServiceNamespace string
+
+	Cfg     *rest.Config
+	RootCAs *x509.CertPool
+}
+
+// Upload uploads the contents of a bundle to the bundle upload service configured on the
+// BundleUploader.
+//
+// To perform the upload, Upload utilizes a Kubernetes API port-forward to forward a port from
+// the uploader service to the local machine. Once the port has been forwarded, Upload
+// uploads the bundleFS as the content for the bundle named by the provided bundleName.
+//
+// Upload returns a boolean value indicating if the bundle content was modified on the server and
+// an error value that will convey any errors that occurred during the upload.
+//
+// Uploads of content that is identical to the existing bundle's content will not result in an error,
+// which means this function is idempotent. Running this function multiple times with the same input
+// does not result in any change to the cluster state after the initial upload.
+func (bu *BundleUploader) Upload(ctx context.Context, bundleName string, bundleFS fs.FS) (bool, error) {
+	pf, err := NewServicePortForwarder(bu.Cfg, types.NamespacedName{Namespace: bu.UploadServiceNamespace, Name: bu.UploadServiceName}, intstr.FromString("https"))
+	if err != nil {
+		return false, err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
Add a comment that we're explicitly canceling in the goroutine further down below?